### PR TITLE
Add InspectionFindings endpoint to Flotilla API

### PR DESCRIPTION
**Pull request by UsamaEquinorAFK**
Using IsarStepId to add InspectionFindings under Inspections which is owned by Tasks which is owned by MissionRun. 

The above statement is now a lie, they are no longer owned.

### CODE REVIEW

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        private readonly IInspectionService _inspectionService;
+        private readonly ILogger<InspectionFindingsController> _logger;
+        public InspectionFindingsController(
+            ILogger<InspectionFindingsController> logger,
+            IInspectionService inspectionService
+        )
+        {
+            _logger = logger;
+            _inspectionService = inspectionService;
+
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:22]:
```suggestion
```

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        /// <summary>
+        /// Add a new inspection finding - need isarStepId to run this
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:26]:
```suggestion
        /// Associate a new inspection finding with the inspection corresponding to isarStepId
```

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        /// </summary>
+        /// <remarks>
+        /// <para> This query adds a new area to the database </para>
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:29]:
```suggestion
```

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        /// </remarks>
+        [HttpPost("add-findings")]
+        [Authorize(Roles = Role.Admin)]
+        [ProducesResponseType(typeof(AreaResponse), StatusCodes.Status201Created)]
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:33]: `AreaResponse` is not the correct return type for this request. Maybe an `Inspection` or creating an `InspectionFindingResponse` would be better?

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:38]: These status codes should be reflected by the actual status codes that might be returned. Remember that this is the documentation that will be displayed in the swagger and should be correct. 

For example, you return OK on line 48 which is a 200 not 201.

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        public async Task<ActionResult<InspectionFindings>> AddFindings([FromBody] InspectionFindingsQuery inspectionFinding)
+        {
+            _logger.LogInformation("Updating inspection findings for inspection with isarStepId '{inspectionFinding.IsarStepId}'", inspectionFinding.IsarStepId);
```
Comment (aestene) [inline backend/api/Controllers/InspectionFindingsController.cs:41]:
```suggestion
            _logger.LogInformation("Updating inspection findings for inspection with isarStepId '{Id}'", inspectionFinding.IsarStepId);
```

File: backend/api/Services/InspectionService.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                }
+                else
+                {
+                    return null;
+                }
+            }
+            else
+            {
+                return null;
+            }
```
Comment (aestene) [inline backend/api/Services/InspectionService.cs:67]:
```suggestion
            else { return null; }
```

File: backend/api/Services/InspectionService.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                    _context.MissionRuns.Update(missionRun);
+                    // Save the changes to the database
+                    await _context.SaveChangesAsync();
+
+                    return inspection;
+                }
+                else
+                {
+                    return null;
+                }
```
Comment (aestene) [inline backend/api/Services/InspectionService.cs:62]:
```suggestion
                else { return null; }
```

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 1 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:22]: I believe this is addressed now, hence resolving.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 3 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: I believe this is addressed now, hence resolving.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 4 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: I believe this is addressed now, hence resolving.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 6 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:40]: I believe this is addressed now, hence resolving.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:40]: Maybe not, why are we doing this ?

File: backend/api/Services/InspectionService.cs (hunk 7 above)
Comment (UsamaEquinorAFK) [reply backend/api/Services/InspectionService.cs]: I believe this is addressed now, hence resolving.

File: backend/api/Services/InspectionService.cs (hunk 8 above)
Comment (UsamaEquinorAFK) [reply backend/api/Services/InspectionService.cs]: I believe this is addressed now, hence resolving.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 2 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: I believe this is addressed now, hence resolving.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 6 above)
Comment (aestene) [reply backend/api/Controllers/InspectionFindingsController.cs:40]: We are doing it for readability. There is not need to specify the entire thing within the string as it is automatically associated with the parameter you put afterwards. Therefore just `Id` or `IsarStepId` is a more suitable name.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Fixed now. Hence resolving.

File: backend/api/Database/Models/InspectionFindings.cs [hunk 9]
```diff
@@ -1,22 +1,46 @@
...
+using Api.Controllers.Models;
 #pragma warning disable CS8618
 namespace Api.Database.Models
 {
-    [Owned]
     public class InspectionFindings
     {
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public string Id { get; set; }
```
Comment (Christdej) [inline backend/api/Database/Models/InspectionFindings.cs:11]: In the Inspection models, I think we should have a reference to this such that an Inspection can contain a List<InspectionFindings>

File: backend/api/Database/Models/InspectionFindings.cs [hunk 10]
```diff
@@ -1,22 +1,46 @@
...
 
-        public string InspectionId { get; set; }
-
-        public string FindingsTag { get; set; }
-
+        public string IsarStepId { get; set; }
+
+        public string Findings { get; set; }
+
+        public string MissionRunId { get; set; }
```
Comment (Christdej) [inline backend/api/Database/Models/InspectionFindings.cs:23]: If we have a link from Inspection, I dont think we need MissionRunId

File: backend/api/Program.cs [hunk 11]
```diff
@@ -73,7 +73,7 @@
 builder.Services.AddScoped<ITaskDurationService, TaskDurationService>();
 builder.Services.AddScoped<IPoseTimeseriesService, PoseTimeseriesService>();
 builder.Services.AddScoped<ILastMissionRunService, LastMissionRunService>();
-
+builder.Services.AddScoped<IInspectionService, InspectionService>();
```
Comment (Christdej) [inline backend/api/Program.cs:76]: This should be InspectionFindingsService right?

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        /// </remarks>
+        [HttpPost("add-findings")]
+        [Authorize(Roles = Role.Admin)]
+        [ProducesResponseType(typeof(InspectionFindingsResponse), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public async Task<ActionResult<InspectionFindings>> AddFindings([FromBody] InspectionFindingsQuery inspectionFinding)
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:38]: Lines 32 says it produces output of type InspectionFindingsResponse while the return type on line 38 says it returns InspectionFindings, this is incorrect yeah?

**Review by Christdej — CHANGES_REQUESTED** [review]

File: backend/api/Program.cs (hunk 11 above)
Comment (UsamaEquinorAFK) [reply backend/api/Program.cs:76]: No we are using inspection service for both Inspection and InspectionFindings so far, as we only need to have an inspection finding when we have an inspection the current approach requires less code and not an additional file.

File: backend/api/Database/Models/InspectionFindings.cs (hunk 10 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/InspectionFindings.cs]: Removed, it was added earlier when we have owned entities. We dont need it anymore.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 12 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Yes it was incorrect. Fixed now I believe.

File: backend/api/Database/Models/InspectionFindings.cs (hunk 9 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/InspectionFindings.cs:11]: But inspection model already contains an inspectionFindingsList

File: backend/api/Program.cs (hunk 11 above)
Comment (UsamaEquinorAFK) [reply backend/api/Program.cs]: Removed redundant line, hence resolved.

File: backend/api/Database/Models/InspectionFindings.cs (hunk 9 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/InspectionFindings.cs:11]: Agreed that we have them as list.

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public async Task<ActionResult<InspectionFindings>> GetInspections([FromRoute] string id)
+        {
+            _logger.LogInformation("Get inspection by ID '{inspectionFinding.InspectionId}'", id);
+            try
+            {
+                var inspection = await _inspectionService.ReadByIsarStepId(id);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:78]: This returns an "Inspection" and in the ProducesResponse and return type of function you say "AreaResponse" and "InspectionFindings", probably a mistake?

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                if (inspection != null)
+                {
+                    return Ok(inspection);
+                }
+
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Error while finding an inspection with inspection id '{id}'", id);
+                throw;
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:88]: Should return InternalServerError or similar

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            return NotFound();
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:90]: Probably add a message to the NotFound

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public async Task<ActionResult<InspectionFindings>> AddFindings([FromBody] InspectionFindingsQuery inspectionFinding)
+        {
+            _logger.LogInformation("Updating inspection findings for inspection with isarStepId '{Id}'", inspectionFinding.IsarStepId);
+            try
+            {
+                var inspection = await _inspectionService.ReadByIsarStepId(inspectionFinding.IsarStepId);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:43]: This returns an "Inspection" and not an "InspectionFindings" as stated in the return type, should perhaps return the list of inspectionfindings on a isarStepId

**Review by Christdej — CHANGES_REQUESTED** [review]

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 13 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:78]: Yup my mistake, Fixed now. Thanks.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:79]: Done. Resolved.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 14 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Done now.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 15 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Done.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 16 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:43]: No, this is for fetching the inspection object based on the isarStepId and then we add inspectionfindings to this object.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:43]: I think it should return inspection as it is required in the next line of code.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:43]: The check on the next line ois based on the premise that there should not be a null inspection to avoid adding any Rogue inspection findings which are not associated with any inspections.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:43]: Done now. Hence resolving. We agreed that we return the finding for the post call and inspection for the getcall.

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+                if (inspection != null)
+                {
+                    inspection = await _inspectionService.AddFindings(inspectionFinding);
+                    return Ok(inspection?.InspectionFindings);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:47]:
```suggestion
                    return Ok(inspection.InspectionFindings);
```
No need for a question mark here I think, as we have checked that inspection is not null.

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+                }
+
+            }
+            catch (Exception e)
+            {
+                const string Message = "Error while adding findings to inspection with IsarStepId '{Id}'";
+                _logger.LogError(e, "Message", inspectionFinding.IsarStepId);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:54]: Will this not just log "Message"?

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+                return StatusCode(StatusCodes.Status500InternalServerError, Message);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:55]: Dont think we have to include this message to the user

File: backend/api/Controllers/InspectionFindingsController.cs [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+                if (inspection != null)
+                {
+                    return Ok(inspection);
+                }
+
+            }
+            catch (Exception e)
+            {
+                const string Message = "Error while finding an inspection with inspection id '{id}'";
+                _logger.LogError(e, "Message", id);
```
Comment (Christdej) [inline backend/api/Controllers/InspectionFindingsController.cs:89]: This will also just log "Message", look at other controllers to see how to log a message with a variable

File: backend/api/Database/Models/InspectionFindings.cs [hunk 21]
```diff
@@ -1,22 +1,42 @@
...
 namespace Api.Database.Models
 {
-    [Owned]
     public class InspectionFindings
     {
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public string Id { get; set; }
 
         public string RobotName { get; set; }
```
Comment (Christdej) [inline backend/api/Database/Models/InspectionFindings.cs:13]: Not sure we need a string RobotName here. The inspection, task, missionRun already has information about which robot ran the mission I think.

File: backend/api/Database/Models/InspectionFindings.cs [hunk 22]
```diff
@@ -1,22 +1,42 @@
...
 
         public string InspectionDate { get; set; }
```
Comment (Christdej) [inline backend/api/Database/Models/InspectionFindings.cs:15]: Should perhaps be a DateTime object

**Review by Christdej — CHANGES_REQUESTED** [review]

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 17 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:47]: Nope it complains and throws a warning otherwise.

Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs:47]: I will see again.

Comment (Christdej) [reply backend/api/Controllers/InspectionFindingsController.cs:47]: Because you already have inspection on line 43, then you override it on line 46

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 20 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Yes my bad. Changed now.

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 18 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: Yes it will, now fixed.

File: backend/api/Database/Models/InspectionFindings.cs (hunk 21 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/InspectionFindings.cs]: Done.

File: backend/api/Database/Models/InspectionFindings.cs (hunk 22 above)
Comment (UsamaEquinorAFK) [reply backend/api/Database/Models/InspectionFindings.cs:13]: Yes I think it should be. Also highlight the correct parameter because this got me thinking why should I have "ID" as date time object :P . Now the InspectionDate is datetime object. =) .

File: backend/api/Controllers/InspectionFindingsController.cs (hunk 17 above)
Comment (UsamaEquinorAFK) [reply backend/api/Controllers/InspectionFindingsController.cs]: This is handled in the lattest ammends.

File: backend/api/Services/InspectionService.cs [hunk 23]
```diff
@@ -61,5 +66,36 @@ private IQueryable<Inspection> GetInspections()
         {
             return _context.Inspections.Include(inspection => inspection.InspectionFindings);
         }
+
+        public async Task<MissionRun?> GetMissionRunFromIsarStepId(string isarStepId)
+        {
+            return await _context.MissionRuns.Where(m => m.Tasks.SelectMany(t => t.Inspections)
+            .Any(i => i.IsarStepId == isarStepId))
+            .FirstOrDefaultAsync();
+        }
```
Comment (Christdej) [inline backend/api/Services/InspectionService.cs:75]: I cannot see where we use this function?

File: backend/api/Migrations/20231109194856_MakeInspectionNotOwned.cs
Comment (Christdej) [inline backend/api/Migrations/20231109194856_MakeInspectionNotOwned.cs:1]: Why are these files changed?

File: backend/api/Migrations/20231109195524_MakeMissionTaskNotOwned.cs
Comment (Christdej) [inline backend/api/Migrations/20231109195524_MakeMissionTaskNotOwned.cs:1]: Why are these files changed?

**Review by Christdej — CHANGES_REQUESTED** [review]

**Review by Christdej — CHANGES_REQUESTED** [review]
Fix the previous comments

File: backend/api/Services/InspectionService.cs (hunk 23 above)
Comment (UsamaEquinorAFK) [reply backend/api/Services/InspectionService.cs:75]: Removed, redundant now.

File: backend/api/Migrations/20231109194856_MakeInspectionNotOwned.cs
Comment (UsamaEquinorAFK) [reply backend/api/Migrations/20231109194856_MakeInspectionNotOwned.cs:1]: not changed now

File: backend/api/Migrations/20231109195524_MakeMissionTaskNotOwned.cs
Comment (UsamaEquinorAFK) [reply backend/api/Migrations/20231109195524_MakeMissionTaskNotOwned.cs:1]: not changed now

**Review by Christdej — APPROVED** [review]
LGTM

**Comment by UsamaEquinorAFK** [comment]
Just rebased afterwards now new work added.
